# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [105]:
import pandas as pd
from utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [106]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [107]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [108]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [109]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [110]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [111]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [112]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [113]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [114]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [115]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [116]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [117]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [118]:
# Write your code here
df["Season"].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [119]:
# Write your code here
# We ese value_counts on the Wteam column to count wins for each team
win_counts = df['Wteam'].value_counts()

# Find the team with the most wins
team_with_most_wins = win_counts.idxmax()
most_wins = win_counts.max()

#We print all the wins to get an overview
print(win_counts)

#This is for which id number that has the highest wins
print(team_with_most_wins)
#This is how many wins the highest id number had
print(most_wins)


Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64
1181
819


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [120]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [121]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [122]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [123]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [124]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [125]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [126]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [127]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [128]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [129]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [130]:
# Write your code here
# Sorting DataFrames
import pandas as pd


# Assuming df is your DataFrame

# We create sorted DataFrames for them all 3
df_sorted_by_season = df.sort_values(by='Season')
df_sorted_by_wteam = df.sort_values(by='Wteam')
df_sorted_by_wscore = df.sort_values(by='Wscore')

# Here we select rows from index 100 to 200 and the columns for season, winning team, and winning score
subset_by_season = df_sorted_by_season.iloc[100:201][['Season', 'Wteam', 'Wscore']]
subset_by_wteam = df_sorted_by_wteam.iloc[100:201][['Season', 'Wteam', 'Wscore']]
subset_by_wscore = df_sorted_by_wscore.iloc[100:201][['Season', 'Wteam', 'Wscore']]





**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [131]:
# Write your code here


# Now we can find the game with the highest winning score
max_wscore_index = subset_by_wscore['Wscore'].idxmax()
max_wscore_row = subset_by_wscore.loc[max_wscore_index]

# Extract the season and winning team
season_max_wscore = max_wscore_row['Season']
winning_team_max_wscore = max_wscore_row['Wteam']


# Print the results
print(f"Season with the highest winning score is: {season_max_wscore}")
print(f"Winning team with the highest winning score is: {winning_team_max_wscore}")






Season with the highest winning score is: 2006
Winning team with the highest winning score is: 1416


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [132]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [133]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [134]:
# Write your code here
# Write your code here

import pandas as pd

# We create the 'ScoreDifference' column
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Now we calculate the average 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()

# and now we filter the DataFrame to only include games with 'ScoreDifference' greater than the average
filtered_df = df[df['ScoreDifference'] > average_score_difference]

# We display the filtered DataFrame
filtered_df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [135]:
# Write your code here

# We find the index of the row with the highest 'ScoreDifference'
max_score_difference_index = filtered_df['ScoreDifference'].idxmax()

# Now we can locate the corresponding row
game_with_highest_difference = filtered_df.loc[max_score_difference_index]

# Extract the season, winning team, and losing team
season_highest_difference = game_with_highest_difference['Season']
winning_team_highest_difference = game_with_highest_difference['Wteam']
losing_team_highest_difference = game_with_highest_difference['Lteam']

# Now we can print the results
print(f"Season with the highest ScoreDifference: {season_highest_difference}")
print(f"Winning team with the highest ScoreDifference: {winning_team_highest_difference}")
print(f"Losing team with the highest ScoreDifference: {losing_team_highest_difference}")


Season with the highest ScoreDifference: 1996
Winning team with the highest ScoreDifference: 1409
Losing team with the highest ScoreDifference: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [136]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [137]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [138]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [139]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [140]:
# Write your code here
# We will group the DataFrame by season and calculate the average winning score for each season
average_winning_score_per_season = df.groupby('Season')['Wscore'].mean()

# Then print the result
print(average_winning_score_per_season)

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [141]:
# Write your code here
# Group the DataFrame by winning team and find the maximum winning score for each team across all seasons
max_winning_score_per_team = df.groupby('Wteam').apply(lambda x: x.loc[x['Wscore'].idxmax()])

# Print the result
print(max_winning_score_per_team[['Season', 'Wscore']])

       Season  Wscore
Wteam                
1101     2015      95
1102     1999     111
1103     1989     109
1104     2005     114
1105     2002     114
...       ...     ...
1460     1993     136
1461     2003     112
1462     1988     125
1463     1987     105
1464     1987     115

[364 rows x 2 columns]


C:\Users\Yusuf\AppData\Local\Temp\ipykernel_22324\1401292779.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  max_winning_score_per_team = df.groupby('Wteam').apply(lambda x: x.loc[x['Wscore'].idxmax()])


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [142]:
# Write your code here

# Then group the DataFrame by both season and winning team, calculate the average winning score, and find the team with the highest average winning score for each season
team_with_highest_avg_score_per_season = df.groupby(['Season', 'Wteam'])['Wscore'].mean().reset_index().groupby('Season').apply(lambda x: x.loc[x['Wscore'].idxmax()])

# Then print the result
print(team_with_highest_avg_score_per_season[['Wteam', 'Wscore']])

         Wteam      Wscore
Season                    
1985    1328.0   92.800000
1986    1109.0   91.200000
1987    1380.0   95.875000
1988    1258.0  111.750000
1989    1258.0  117.315789
1990    1258.0  126.347826
1991    1380.0  112.312500
1992    1380.0   99.642857
1993    1380.0  101.875000
1994    1380.0  106.583333
1995    1206.0  102.833333
1996    1206.0  103.600000
1997    1254.0   94.142857
1998    1395.0  101.153846
1999    1317.0   95.750000
2000    1395.0   95.647059
2001    1395.0   99.315789
2002    1242.0   92.888889
2003    1395.0   88.222222
2004    1194.0   92.142857
2005    1311.0   91.333333
2006    1190.0   89.533333
2007    1440.0   99.400000
2008    1440.0   94.500000
2009    1377.0   96.200000
2010    1440.0   95.000000
2011    1449.0   90.130435
2012    1370.0   87.400000
2013    1377.0   86.500000
2014    1322.0   92.600000
2015    1322.0   90.733333
2016    1146.0   97.142857


C:\Users\Yusuf\AppData\Local\Temp\ipykernel_22324\1324789433.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  team_with_highest_avg_score_per_season = df.groupby(['Season', 'Wteam'])['Wscore'].mean().reset_index().groupby('Season').apply(lambda x: x.loc[x['Wscore'].idxmax()])


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [143]:
# Write your code here
# Group the DataFrame by both season and winning team and count the number of wins
wins_per_team_per_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Print the result
print(wins_per_team_per_season)


       Season  Wteam  Wins
0        1985   1102     5
1        1985   1103     9
2        1985   1104    21
3        1985   1106    10
4        1985   1108    19
...       ...    ...   ...
10167    2016   1460    20
10168    2016   1461    12
10169    2016   1462    27
10170    2016   1463    21
10171    2016   1464     9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [144]:
# Write your code here

# We group the DataFrame by both season and winning team and count the number of wins
wins_per_team_per_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Now find the index of the team with the most wins for each season
index_of_most_wins_per_season = wins_per_team_per_season.groupby('Season')['Wins'].idxmax()

# We can now get the team with the most wins for each season
team_with_most_wins_per_season = wins_per_team_per_season.loc[index_of_most_wins_per_season]

# We print the result
print(team_with_most_wins_per_season)


      Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [145]:
# Write your code here
# We group the DataFrame by losing team and find the average losing score for each team across all seasons
average_losing_score_per_team = df.groupby('Lteam')['Lscore'].mean()

# We will now compare the average losing score with the average winning score for each team, to see the difference 
for team in df['Wteam'].unique():
    if team in average_losing_score_per_team.index:
        avg_winning_score = df[df['Wteam'] == team]['Wscore'].mean()
        avg_losing_score = average_losing_score_per_team[team]
        
        print(f"Team: {team}")
        print(f"Average Winning Score: {avg_winning_score}")
        print(f"Average Losing Score: {avg_losing_score}")
        
        if avg_losing_score > avg_winning_score:
            print("This team has a higher average losing score than winning score.")
        else:
            print("This team does not have a higher average losing score than winning score.")
        print()


Team: 1228
Average Winning Score: 77.56666666666666
Average Losing Score: 63.4789156626506
This team does not have a higher average losing score than winning score.

Team: 1106
Average Winning Score: 76.63451776649747
Average Losing Score: 66.28416485900217
This team does not have a higher average losing score than winning score.

Team: 1112
Average Winning Score: 82.21983914209115
Average Losing Score: 70.05531914893616
This team does not have a higher average losing score than winning score.

Team: 1165
Average Winning Score: 73.29131652661064
Average Losing Score: 61.121848739495796
This team does not have a higher average losing score than winning score.

Team: 1192
Average Winning Score: 76.31529411764706
Average Losing Score: 64.92753623188406
This team does not have a higher average losing score than winning score.

Team: 1218
Average Winning Score: 76.33744855967078
Average Losing Score: 65.96629213483146
This team does not have a higher average losing score than winning score.

Team: 1325
Average Winning Score: 76.26987060998151
Average Losing Score: 65.94763092269326
This team does not have a higher average losing score than winning score.

Team: 1326
Average Winning Score: 77.2379421221865
Average Losing Score: 65.73121387283237
This team does not have a higher average losing score than winning score.

Team: 1328
Average Winning Score: 84.63595839524517
Average Losing Score: 68.63728813559322
This team does not have a higher average losing score than winning score.

Team: 1339
Average Winning Score: 74.58132530120481
Average Losing Score: 65.13528336380256
This team does not have a higher average losing score than winning score.

Team: 1365
Average Winning Score: 73.06521739130434
Average Losing Score: 63.67785234899329
This team does not have a higher average losing score than winning score.

Team: 1375
Average Winning Score: 77.62446351931331
Average Losing Score: 65.17209302325581
This team does not have a higher average losing score than winning score.


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [146]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [147]:
# Write your code here
# Iterate over the rows of the DataFrame and set 'HighScoringGame' based on the winning score
for index, row in df.iterrows():
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'
    else:
        df.at[index, 'HighScoringGame'] = 'No'

# Print the result
print(df[['Wscore', 'HighScoringGame']])

        Wscore HighScoringGame
0           81              No
1           77              No
2           63              No
3           70              No
4           86              No
...        ...             ...
145284      70              No
145285      72              No
145286      82              No
145287      66              No
145288      87              No

[145289 rows x 2 columns]


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [148]:
# Write your code here

# Initialize an empty dictionary to keep track of the number of games for each team
games_played = {}

# Iterate over the rows of the DataFrame and update the dictionary
for index, row in df.iterrows():
    # Update the number of games for the winning team
    if row['Wteam'] in games_played:
        games_played[row['Wteam']] += 1
    else:
        games_played[row['Wteam']] = 1
    
    # Update the number of games for the losing team
    if row['Lteam'] in games_played:
        games_played[row['Lteam']] += 1
    else:
        games_played[row['Lteam']] = 1

# Print the result
for team, num_games in games_played.items():
    print(f"Team: {team}, Total Games Played: {num_games}")


Team: 1228, Total Games Played: 992
Team: 1328, Total Games Played: 968
Team: 1106, Total Games Played: 855
Team: 1354, Total Games Played: 906
Team: 1112, Total Games Played: 981
Team: 1223, Total Games Played: 363
Team: 1165, Total Games Played: 833
Team: 1432, Total Games Played: 69
Team: 1192, Total Games Played: 908
Team: 1447, Total Games Played: 903
Team: 1218, Total Games Played: 931
Team: 1337, Total Games Played: 922
Team: 1226, Total Games Played: 847
Team: 1242, Total Games Played: 993
Team: 1268, Total Games Played: 969
Team: 1260, Total Games Played: 914
Team: 1133, Total Games Played: 949
Team: 1305, Total Games Played: 922
Team: 1424, Total Games Played: 974
Team: 1307, Total Games Played: 969
Team: 1288, Total Games Played: 925
Team: 1344, Total Games Played: 951
Team: 1438, Total Games Played: 952
Team: 1374, Total Games Played: 916
Team: 1411, Total Games Played: 903
Team: 1412, Total Games Played: 962
Team: 1397, Total Games Played: 963
Team: 1417, Total Games Playe

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [149]:
# Write your code here

# Initialize a dictionary to keep track of the highest score difference for each season
highest_score_difference_per_season = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Calculate the score difference for the current row
    score_difference = row['Wscore'] - row['Lscore']
    
    # Update the highest score difference for the current season if necessary
    if row['Season'] in highest_score_difference_per_season:
        if score_difference > highest_score_difference_per_season[row['Season']]['score_difference']:
            highest_score_difference_per_season[row['Season']] = {
                'game_index': index,
                'score_difference': score_difference
            }
    else:
        highest_score_difference_per_season[row['Season']] = {
            'game_index': index,
            'score_difference': score_difference
        }

# Print the result
for season, game_info in highest_score_difference_per_season.items():
    print(f"Season: {season}, Highest Score Difference: {game_info['score_difference']}")
    print(df.loc[game_info['game_index']])


Season: 1985, Highest Score Difference: 60
Season             1985
Daynum               33
Wteam              1361
Wscore              128
Lteam              1288
Lscore               68
Wloc                  H
Numot                 0
ScoreDifference      60
HighScoringGame     Yes
Name: 236, dtype: object
Season: 1986, Highest Score Difference: 84
Season             1986
Daynum               60
Wteam              1314
Wscore              129
Lteam              1264
Lscore               45
Wloc                  N
Numot                 0
ScoreDifference      84
HighScoringGame     Yes
Name: 4731, dtype: object
Season: 1987, Highest Score Difference: 73
Season             1987
Daynum               51
Wteam              1155
Wscore              112
Lteam              1118
Lscore               39
Wloc                  H
Numot                 0
ScoreDifference      73
HighScoringGame     Yes
Name: 8240, dtype: object
Season: 1988, Highest Score Difference: 68
Season             1988
Daynum 

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [150]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [151]:
# Write your code here

# Calculate the total number of games played by each team using vectorized operations
total_games_played = df['Wteam'].value_counts() + df['Lteam'].value_counts()

# Rename the axis to provide column headers
total_games_played = total_games_played.rename_axis('Team').reset_index(name='Total Games Played')

# Print the result
print(total_games_played)


     Team  Total Games Played
0    1101                  76
1    1102                 840
2    1103                 910
3    1104                 975
4    1105                 447
..    ...                 ...
359  1460                 827
360  1461                 914
361  1462                 954
362  1463                 838
363  1464                 856

[364 rows x 2 columns]


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [152]:
# Write your code here

# Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Use groupby() and idxmax() to find the game with the highest score difference for each season
game_with_highest_difference_per_season = df.groupby('Season').apply(lambda x: x.loc[x['ScoreDifference'].idxmax()])

# Print the result
print(game_with_highest_difference_per_season[['Wteam', 'Lteam', 'Wscore', 'Lscore', 'ScoreDifference']])


        Wteam  Lteam  Wscore  Lscore  ScoreDifference
Season                                               
1985     1361   1288     128      68               60
1986     1314   1264     129      45               84
1987     1155   1118     112      39               73
1988     1328   1147     152      84               68
1989     1242   1135     115      45               70
1990     1181   1217     130      54               76
1991     1163   1148     115      47               68
1992     1116   1126     128      46               82
1993     1328   1197     146      65               81
1994     1228   1152     121      52               69
1995     1246   1404     124      50               74
1996     1409   1341     141      50               91
1997     1278   1106     114      34               80
1998     1395   1410     153      87               66
1999     1268   1317     132      57               75
2000     1261   1212     112      37               75
2001     1196   1197     125

C:\Users\Yusuf\AppData\Local\Temp\ipykernel_22324\2960445991.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  game_with_highest_difference_per_season = df.groupby('Season').apply(lambda x: x.loc[x['ScoreDifference'].idxmax()])


# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [153]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [154]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [155]:
type(df['Wscore'])

pandas.core.series.Series

In [156]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [157]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [158]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [159]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y